pan3d/dataset_builder.py

```diff
@@ -199,43 +210,52 @@ def set_dataset_path(self, dataset_path, **kwargs):
                     dataset_path, engine=engine, chunks={}
                 )
             except Exception as e:
-                self.state.error_message = str(e)
+                self.state.ui_error_message = str(e)
                 return
         else:
             # Assume it is a named tutorial dataset
             self.dataset = xarray.tutorial.load_dataset(dataset_path)
-        self.state.data_vars = [
+        # reset algorithm
+        self.algorithm = PyVistaXarraySource()
+        self.state.da_vars = [
             {"name": k, "id": i} for i, k in enumerate(self.dataset.data_vars.keys())
         ]
-        self.state.data_attrs = [
+        self.state.da_attrs = [
             {"key": k, "value": v} for k, v in self.dataset.attrs.items()
         ]
-        self.state.data_attrs.insert(
+        self.state.da_attrs.insert(
             0,
             {
                 "key": "dimensions",
                 "value": str(dict(self.dataset.dims)),
             },
         )
-        if len(self.state.data_attrs) > 0:
-            self.state.show_data_attrs = True
-        self.state.coordinates = []
-        self.state.expanded_coordinates = []
+        self.state.da_coordinates = []
+        self.state.ui_expanded_coordinates = []
         self.state.update(
-            dict(x_array=None, y_array=None, z_array=None, t_array=None, t_index=0)
+            dict(da_x=None, da_y=None, da_z=None, da_t=None, da_t_index=0)
         )
         self.state.dataset_ready = True
-        if len(self.state.data_vars) > 0:
-            self.state.array_active = self.state.data_vars[0]["name"]
+        if len(self.state.da_vars) > 0:
+            self.set_da_active(self.state.da_vars[0]["name"])
         else:
-            self.state.no_data_vars = True
-        self.state.loading = False
-
-    @change("array_active")
-    def on_set_array_active(self, array_active, **kwargs):
-        if array_active is None or not self.state.dataset_ready:
+            self.state.no_da_vars = True
+        self.state.ui_loading = False
+
+    @change("da_active")
+    def set_da_active(self, da_active, **kwargs):
+        if (
+            da_active is None
+            or not self.state.dataset_ready
+            or da_active == self.da_active
+        ):
             return
+        self.da_active = da_active
+        self.state.da_active = da_active
```

Comment:
I don't get that line. This is equivalent to `self.state.da_active = self.state.da_active`

Maybe that is why you called those methods set? 

What is the intent of that method? update_active_array?

Reply:
This was my attempt to mimic the suggestion you had made for `set_dataset_path`, where the method can be either called as a state change effect or called directly with a new value. When called in the latter case, this line is necessary. But a redundant call is avoided with the check just above it. `self.da_active` is stored separately from `self.state.da_active` just like you had suggested for `self.dataset_path` being separate from `self.state.dataset_path`. It is mostly redundant storage, but it allows us to check if the side effect has already been called for that value.

Reply:
That makes sense, especially if that method is not called by the @change. 
Another option (if that makes better sense) would be to do something like

```python
def set_active_array_name(array_name):
    self.state.da_active = array_name

@change("da_active")
def _on_active_array_change(da_active, **kwargs):
    ...
```

Reply:
I do like this separation of a user-accessible functions from the state change side effects. It makes it no longer necessary to have redundant storage like `self.dataset_path` versus `self.state.dataset_path`. I'll apply this pattern to all state changes that should be user-accessible.

Reply:
I am trying this out now, and I'm making an example notebook which uses these user-accessible functions back-to-back, like so:
```
builder = DatasetBuilder()

builder.set_dataset_path('../example_dataset.nc')
builder.set_data_array_active_name('density')
builder.set_data_array_axis_names(
    x='length',
    y='width',
    z='height',
    t='seconds',
)
builder.set_data_array_time_index(2)
```
But this breaks because the state change callbacks are not called in-between. The callback for when `dataset_path` is set needs to occur before the callback after `da_active` is set. Since we cannot enforce callback order on the final flush, Is there a way that we can cause a flush to occur in these user-accessible functions?

Reply:
Probably yes, but I would not advise it. 

Just keep that specific one that need explicit execution order the way you have it.

Reply:
Maybe I can try a combination of the techniques. The user-accessible function could call the private function, and the private function can still use a check against `self.state.dataset_path` equaling `self.dataset_path`? This will avoid performing the side effects twice.